src/poly/dcrt/sampler/hash.rs

```diff
@@ -91,45 +94,41 @@ where
                 let bit_length = params.modulus_bits();
                 let index = (nrow * ncol * n * bit_length).div_ceil(hash_output_size);
                 // bits = number of resulting bits from hashing ops = hash_output_size * index
-                let mut bits = Vec::with_capacity(hash_output_size * index);
-                let mut ring_elems = Vec::with_capacity((index * hash_output_size) / bit_length);
-                for i in 0..(index) {
+                let mut bv = bitvec![u8, Msb0;];
+                let mut og_hasher: H = H::new();
+                og_hasher.update(self.key);
+                og_hasher.update(tag.as_ref());
+                info!("before loop {}, {}", index, bit_length);
+                for i in 0..index {
+                    let mut hasher = og_hasher.clone();
                     //  H ( key || tag || i )
-                    let mut hasher = H::new();
-                    // todo: we currently assuming index is less than u32
-                    let min_i_type = std::mem::size_of_val(&i);
-                    if min_i_type > std::mem::size_of::<u8>() {
-                        let mut combined =
-                            Vec::with_capacity(self.key.len() + tag.as_ref().len() + 32);
-                        combined.extend_from_slice(&self.key);
-                        combined.extend_from_slice(tag.as_ref());
-                        combined.extend_from_slice(&i.to_be_bytes());
-                        hasher.update(&combined);
-                    } else {
-                        let mut combined =
-                            Vec::with_capacity(self.key.len() + tag.as_ref().len() + 1);
-                        combined.extend_from_slice(&self.key);
-                        combined.extend_from_slice(tag.as_ref());
-                        combined.push(i as u8);
-                        hasher.update(&combined);
-                    }
-
+                    hasher.update(i.to_be_bytes());
                     for &byte in hasher.finalize().iter() {
-                        for bit_index in 0..8 {
-                            let bit = (byte >> bit_index) & 1;
-                            bits.push(bit);
+                        for bit_index in (0..8).rev() {
+                            bv.push((byte >> bit_index) & 1 != 0);
```

Comment:
@rkdud007 Do you have any reason to order bits from the most-significant bit?

Reply:
good point changing into LSB